File: python/goldenmodel.py
import numpy as np
from numpy import (abs, log10, min, max)
from numpy import fft
from numpy import random
from scipy import signal
import matplotlib.pyplot as plt

def golden(x, h, yi, M, D, decmod, dt=np.complex128):

  L = int(np.ceil((M-decmod)/D))
  Y = np.zeros((M,L), dtype=dt)
  n = np.arange(0, M)

  for k in range(0, M):
    # mix down
    argmix = (-1j*2*np.pi*k*n)/M
    xmix = x*np.exp(argmix)

    # low pass filter
    ymix, yi[k,:] = signal.lfilter(h, 1, xmix, zi=yi[k,:])

    # decimate
    ydec = ymix[decmod:-1:D]

    Y[k,:] = ydec

  decmod = (D-M+decmod) % D

  return (Y, yi, L, decmod)

def pltfine(FineSpectrumMat, fs, NFFT_FINE, M, D):
  fs_os = fs/D
  hsov = (M-D)*NFFT_FINE//(2*M)
  
  fig, ax = plt.subplots(4,8, sharey='row')
  for i in range(0,4):
    for j in range(0,8):
      k = (i*8)+j
      # this shift corrects for overlap between adjacent bins
      bin_shift = - ((NFFT_FINE//2) + k*2*hsov)

      subbins = np.arange(k*NFFT_FINE, (k+1)*NFFT_FINE) + bin_shift
      cur_ax = ax[i,j]
      cur_ax.plot(subbins*fs_os/NFFT_FINE, 20*log10(abs(FineSpectrumMat[k, :])))
      cur_ax.set_xlim(min(subbins*fs_os/NFFT_FINE), max(subbins*fs_os/NFFT_FINE))
      cur_ax.grid(True)
  plt.show()

def crandn(shape):
  return (random.randn(shape) + 1j*random.randn(shape))

if __name__=="__main__":
  print("********* BANK OF DECIMATED LPF ************")

  # simulation parameters
  M = 32
  D = 24
  P = 8
  L = M*P

  osratio = M/D

  fs = 10e3
  f_soi = 2e3

  NFFT = M
  NFFT_FINE = 512

  FINE_FRAMES = 10

  hsov = (M-D)*NFFT_FINE//(2*M)

  # prototype LPF filter taps
  sincarg = np.arange(-P/2*osratio, osratio*P/2, 1/D)
  sinc = np.sinc(sincarg)
  hann = np.hanning(L)
  h = sinc*hann

  # shaping filter
  blueNoise = fft.ifft(fft.fftshift(np.sqrt(np.linspace(1,3,1024)))) # TODO: some magic numbers here
  zi = np.zeros(1024-1, dtype=np.complex128)

  X = np.zeros((NFFT, NFFT_FINE), dtype=np.complex128)
  Xf = np.zeros((NFFT*NFFT_FINE, FINE_FRAMES), dtype=np.complex128)
  Xfpruned = np.zeros((D*NFFT_FINE, FINE_FRAMES), dtype=np.complex128)

  yi = np.zeros((M, L-1), dtype=np.complex128)
  n = np.arange(0,M)
  nn = 0
  decmod = 0

  for i in range(0, FINE_FRAMES):
    st = 0
    ed = 0
    X = np.zeros((NFFT, NFFT_FINE), dtype=np.complex128)

    while ed < NFFT_FINE:
      # generate filterd noiese samples 
      argf = (1j*2*np.pi*f_soi/fs*(np.arange(nn*M, (nn+1)*M)))
      nn += 1
      sig = np.exp(argf)
      noise = crandn(M)
      xh = sig + noise
      #xh = crandn(M)
      #xh, zi = signal.lfilter(blueNoise, 1, x, zi=zi)

      (Ydec, yi, ndec, decmod) = golden(xh, h, yi, M, D, decmod)

      ed = st + ndec
      X[:, st:ed] = Ydec
      st = ed

    # compute a the fine spectrum output along with the corrected with
    # oversample regions discarded
    fine = fft.fftshift(fft.fft(X, NFFT_FINE, axis=1), axes=(1,))/NFFT_FINE
    finepruned = fine[:,(hsov-1):-(hsov+1)]
    Xf[:, i] = fine.reshape(NFFT*NFFT_FINE)
    Xfpruned[:,i] = finepruned.reshape(D*NFFT_FINE)

  # plot averaged fine spectrum
  Sxx_pruned = np.mean(np.real(Xfpruned*np.conj(Xfpruned)), 1)
  fshift = -(NFFT_FINE/2-hsov+1)
  fbins_fine = np.arange(0, D*NFFT_FINE) + fshift
  faxis_fine = fbins_fine*fs/D/NFFT_FINE

  plt.plot(faxis_fine, 10*np.log10(Sxx_pruned))
  plt.ylim([-20, 70])
  plt.grid()
  plt.show()

File: mat/goldenmodel.m
% bank of deimated FIR filters
clearvars;
M = 32;
D = 24;
P=8;

L = M*P;

% Design prototype LPF (Hanning window)
idx = -P/2*(M/D):1/D:(M/D)*P/2-1/D; % (M/D) 1/D to extend to the correct L length. Note: need to design the filter for the decimation rate D not the maximally sampled rate M in fact I really need to investiagate proper filter design here to get a super tight corner to achieve an image rejection as low as askap does of -60dB.
x = sinc(idx);
hann = hanning(L).';
h = x.*hann;

% h = fir1(L-1, 0.95/D,'low',hamming(L));

osratio = M/D;


fs = 10e3;
f_soi = 2000;

Nfft = M;
Nfft_fine = 512;
hsov = (M-D)*Nfft_fine/(2*M);

fine_frames = 10;

X = zeros(Nfft, Nfft_fine);
Xf = zeros(Nfft*Nfft_fine, fine_frames);
Xfpruned = zeros(D*Nfft_fine, fine_frames);

yi = zeros(M,L-1);
n = (0:M-1)';
decmod = 0; % variable to track where to start from in next decimation sequence
nn = 0;
%%
for i=1:fine_frames
  st = 1;
  ed = 1;
  X = zeros(Nfft, Nfft_fine);
  while ed <= Nfft_fine
    % generate a sequence of input data
    argf = (1j*2*pi*f_soi/fs*(nn*M:(nn+1)*M-1)).';
    nn = nn+1;
    sig = exp(argf);
    noise = randn(M,1) + 1j*randn(M,1);
    x = sig + noise;
    
    for k = 1:M
      % mix
      xmix = x.*exp(-1j*2*pi*(k-1)*n/M);
      % lpf
      [ymix, yi(k,:)] = filter(h, 1, xmix, yi(k,:));
      % decimate
      ydec = ymix(1+decmod:D:end);
      % store
      ed = st+length(ydec);
      X(k, st:ed-1) = ydec;
    end
    st = ed;
    decmod = mod(D-M+decmod,D);
  end
  fine = fftshift(fft(X, Nfft_fine,2),2)/Nfft_fine; % also need to fftshift on the 2 dim here...
  fine_pruned = fine(:, hsov:end-hsov-1); % which is it... (hsov+1:end-hsov) or... ... it looks like the current one is correct
  Xf(:, i) = reshape(fine.', [Nfft*Nfft_fine,1]); % the .' on the reshape always gets me...
  Xfpruned(:, i) = reshape(fine_pruned.', [D*Nfft_fine,1]); % the .' on the reshape always gets me...
end
%%

fs_os = fs/D;
fshift = -(Nfft_fine/2-hsov+1);
fbins = (0:M*Nfft_fine-1)*fs/M/Nfft_fine;
fbins_pruned = ((0:D*Nfft_fine-1) + fshift)*fs_os/Nfft_fine;
% 
% figure(1);
% plot(fbins_pruned, 20*log10(abs(Xfpruned(:,1))))
% xlim([min(fbins_pruned), max(fbins_pruned)]);

% figure(2);
% Sxx = mean(abs(Xf).^2,2);
% plot(fbins, 10*log10(Sxx));

figure(3);
Sxx_pruned = mean(real(Xfpruned.*conj(Xfpruned)), 2);
plot(fbins_pruned, 10*log10(Sxx_pruned)); grid on;
xlim([min(fbins_pruned), max(fbins_pruned)]);
% ylim([-20, 70]);
% 
% for m = 1:M
%   figure(4);
%   subplot(4,8,m);
%   fbins_corrected = ((m-1)-1/2)*(Nfft_fine-2*hsov):((m-1)+1/2)*(Nfft_fine-2*hsov)-1;
%   f_corrected = fbins_corrected*fs_os/Nfft_fine;
%   plot(f_corrected, 20*log10(abs(fine_pruned(m,:)))); grid on;
%   xlim([min(f_corrected), max(f_corrected)]);
%   ylim([-60, 40]);
% end
% 
% % faxis not really correct here because it is linear. True faxis would
% % share edges with adjacent axes.
% for m = 1:M
%   figure(5);
%   subplot(4,8,m);
%   fbins_corrected = ((m-1)-1/2)*(Nfft_fine):((m-1)+1/2)*Nfft_fine-1;
%   f_corrected = fbins_corrected*fs_os/Nfft_fine;
%   plot(f_corrected, 20*log10(abs(fine(m,:)))); grid on;
%   xlim([min(f_corrected), max(f_corrected)]);
%   ylim([-60, 40]);
% end
